library/kotlin/src/io/envoyproxy/envoymobile/StatsClientImpl.kt

```diff
@@ -0,0 +1,16 @@
+package io.envoyproxy.envoymobile
+
+import io.envoyproxy.envoymobile.engine.EnvoyEngine
+import java.lang.ref.WeakReference
+
+/**
+ * Envoy implementation of `StatsClient`.
+ */
+internal class StatsClientImpl constructor(
+  internal val engine: EnvoyEngine
+) : StatsClient {
+
+  override fun getCounter(vararg elements: Element): Counter {
+    return Counter(WeakReference(engine), elements.asList())
```

Comment:
iOS has a strong ref below, we should keep them consistent (I think a strong ref is what you want here)

Reply:
Ah I see, this is passing a weak reference to the Counter, whereas on iOS the Counter itself makes the reference weak. Would it make sense to move `WeakReference(engine)` into the counter file's constructor to make it clear that it doesn't retain the engine?

Separately, I'm wondering if it makes sense to allow counters to retain the engine. Do we want the engine to stick around as long as a consumer is holding on to a counter ref?

Reply:
> Would it make sense to move WeakReference(engine) into the counter file's constructor to make it clear that it doesn't retain the engine?

yea

> Do we want the engine to stick around as long as a consumer is holding on to a counter ref?

probably not, counters could be referenced by feature modules of the platform(iOS/android) code, and we probably don't want the feature modules to (indirectly) hold a strong ref to engine

Reply:
My intuition right now is that it seems natural for engine *clients* to strongly retain the engine, but perhaps less so for the artifacts they produce. This has more to do with users of the library being able to implicitly manage the lifecycle of the engine than anything else. I should say that I'm not _strongly_ convinced this is right, but it seems reasonable for now for experimentation.

Reply:
Was there a resolution here? Looks like the weak ref is still being passed from this line